### DISCUSSION TITLE
'NoneType' object has no attribute 'choices'

### DISCUSSION THREAD

**Discussion by abhijitsharma1010** [netbox-community/netbox#13654] [Help Wanted!]
![12](https://github.com/netbox-community/netbox/assets/44794851/0120bf43-28c6-4cc6-9867-7beea4c73099)

After upgrading to v3.6.0 , when we click on virtual machine tab from the side menu, we get this error "'NoneType' object has no attribute 'choices' " same thing happens when we try to go in a Vm from change log and click on edit button. 

Any help to resolve this will be very much appreciated.

**Comment by candlerb** [comment]
Works for me, so there's something about your system that's different. Do you have any customizations, e.g. custom fields on VirtualMachine?

Can you provide a backtrace? (Set DEBUG=True in configuration.py, or set ADMINS and EMAIL to have it automatically mailed to you whenever an exception occurs in production)

**Reply by abhijitsharma1010** [reply]
Environment:


Request Method: GET
Request URL: https://10.40.122.145/virtualization/virtual-machines/

Django Version: 4.2.4
Python Version: 3.8.10
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.humanize',
 'corsheaders',
 'debug_toolbar',
 'graphiql_debug_toolbar',
 'django_filters',
 'django_tables2',
 'django_prometheus',
 'graphene_django',
 'mptt',
 'rest_framework',
 'social_django',
 'taggit',
 'timezone_field',
 'core',
 'account',
 'circuits',
 'dcim',
 'ipam',
 'extras',
 'tenancy',
 'users',
 'utilities',
 'virtualization',
 'wireless',
 'django_rq',
 'drf_spectacular',
 'drf_spectacular_sidecar']
Installed Middleware:
['graphiql_debug_toolbar.middleware.DebugToolbarMiddleware',
 'django_prometheus.middleware.PrometheusBeforeMiddleware',
 'corsheaders.middleware.CorsMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'netbox.middleware.RemoteUserMiddleware',
 'netbox.middleware.CoreMiddleware',
 'netbox.middleware.MaintenanceModeMiddleware',
 'django_prometheus.middleware.PrometheusAfterMiddleware']



Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/base.py", line 77, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/utilities/views.py", line 99, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/bulk_views.py", line 179, in get
    'filter_form': self.filterset_form(request.GET, label_suffix='') if self.filterset_form else None,
  File "/opt/netbox/netbox/netbox/forms/base.py", line 166, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/netbox/netbox/utilities/forms/mixins.py", line 19, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/netbox/netbox/extras/forms/mixins.py", line 30, in __init__
    self._append_customfield_fields()
  File "/opt/netbox/netbox/extras/forms/mixins.py", line 57, in _append_customfield_fields
    self.fields[field_name] = self._get_form_field(customfield)
  File "/opt/netbox/netbox/netbox/forms/base.py", line 181, in _get_form_field
    return customfield.to_form_field(set_initial=False, enforce_required=False, enforce_visibility=False)
  File "/opt/netbox/netbox/extras/models/customfields.py", line 434, in to_form_field
    choices = self.choice_set.choices

Exception Type: AttributeError at /virtualization/virtual-machines/
Exception Value: 'NoneType' object has no attribute 'choices'


I hope this will help in troubleshooting.

**Reply by abhijitsharma1010** [reply]
Yes we are using  some custom fields also in Virtual machines.

**Reply by candlerb** [reply]
It looks like you have a custom field of type "select" or "multiselect", but it doesn't have any choice set assigned.

Did the migrations not complete successfully when you ran `./upgrade.sh` ?

Check Customization > Custom Field Choices and Customization > Custom Fields; each one of type "Selection" or "Multiple Selection" should have a choice set assigned.

You could fix this by hand, but I'd be more worried that you're missing other migrations.

**Reply by candlerb** [reply]
I can suggest a small change which might mask this problem.

Try changing [line 434](https://github.com/netbox-community/netbox/blob/v3.6.0/netbox/extras/models/customfields.py#L434) of `/opt/netbox/netbox/extras/models/customfields.py` from

```
            choices = self.choice_set.choices
```

to

```
            choices = self.choices
```

(and restart netbox)

That's because there is a 'choices' property on CustomField defined [here](https://github.com/netbox-community/netbox/blob/v3.6.0/netbox/extras/models/customfields.py#L228-L232):

```
    @property
    def choices(self):
        if self.choice_set:
            return self.choice_set.choices
        return []
```

This should give you a valid (but empty) choice set even if choice_set is None.

If this fixes the problem, it could be raised as a PR. But I'd still want to understand how your database got into its current broken state.

**Reply by abhijitsharma1010** [reply]
I tried the changes mentioned above, but still get this error.
![13](https://github.com/netbox-community/netbox/assets/44794851/2565e47f-2864-49bd-a220-4255241e8f6a)

**Reply by abhijitsharma1010** [reply]
Also attaching the cli output from the # ./upgrade.sh
[upgrade logs.txt](https://github.com/netbox-community/netbox/files/12511648/upgrade.logs.txt)

**Reply by candlerb** [reply]
Backtrace for the new error please?

**Reply by abhijitsharma1010** [reply]
Environment:


Request Method: GET
Request URL: https://10.40.122.145/virtualization/virtual-machines/

Django Version: 4.2.4
Python Version: 3.8.10
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.humanize',
 'corsheaders',
 'debug_toolbar',
 'graphiql_debug_toolbar',
 'django_filters',
 'django_tables2',
 'django_prometheus',
 'graphene_django',
 'mptt',
 'rest_framework',
 'social_django',
 'taggit',
 'timezone_field',
 'core',
 'account',
 'circuits',
 'dcim',
 'ipam',
 'extras',
 'tenancy',
 'users',
 'utilities',
 'virtualization',
 'wireless',
 'django_rq',
 'drf_spectacular',
 'drf_spectacular_sidecar']
Installed Middleware:
['graphiql_debug_toolbar.middleware.DebugToolbarMiddleware',
 'django_prometheus.middleware.PrometheusBeforeMiddleware',
 'corsheaders.middleware.CorsMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'netbox.middleware.RemoteUserMiddleware',
 'netbox.middleware.CoreMiddleware',
 'netbox.middleware.MaintenanceModeMiddleware',
 'django_prometheus.middleware.PrometheusAfterMiddleware']



Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/base.py", line 77, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/utilities/views.py", line 99, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/bulk_views.py", line 179, in get
    'filter_form': self.filterset_form(request.GET, label_suffix='') if self.filterset_form else None,
  File "/opt/netbox/netbox/netbox/forms/base.py", line 166, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/netbox/netbox/utilities/forms/mixins.py", line 19, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/netbox/netbox/extras/forms/mixins.py", line 30, in __init__
    self._append_customfield_fields()
  File "/opt/netbox/netbox/extras/forms/mixins.py", line 57, in _append_customfield_fields
    self.fields[field_name] = self._get_form_field(customfield)
  File "/opt/netbox/netbox/netbox/forms/base.py", line 181, in _get_form_field
    return customfield.to_form_field(set_initial=False, enforce_required=False, enforce_visibility=False)
  File "/opt/netbox/netbox/extras/models/customfields.py", line 461, in to_form_field
    widget=widget_class(api_url=f'/api/extras/custom-field-choice-sets/{self.choice_set.pk}/choices/')

Exception Type: AttributeError at /virtualization/virtual-machines/
Exception Value: 'NoneType' object has no attribute 'pk'

**Reply by candlerb** [reply]
Also, could you try these Postgres queries

```
select id,type,name from extras_customfield where type in ('select','multiselect') and choice_set_id is null;

select id,name from extras_customfieldchoiceset;
```

**Reply by abhijitsharma1010** [reply]
netbox@netbox-vm:/opt/netbox/netbox/netbox$ psql
psql (12.16 (Ubuntu 12.16-0ubuntu0.20.04.1))
Type "help" for help.

netbox=> select id,type,name from extras_customfield where type in ('select','multiselect') and choice_set_id is null;
 id |  type  |  name
----+--------+--------
 10 | select | Server
(1 row)

netbox=> select id,name from extras_customfieldchoiceset;
 id |              name
----+---------------------------------
  1 | Third Party Permissions Choices
  2 | ASN Choices
  3 | Network Role Choices
  4 | Procured By Choices
  5 | whois parameters Choices
(5 rows)

**Reply by candlerb** [reply]
OK, that's another case where `self.choice_set` is assumed to be not None.  It would have to be coded around.

**Reply by abhijitsharma1010** [reply]
@candlerb do I need to revert back to a 3.5 series build for the mean time? What do you suggest?

**Reply by candlerb** [reply]
OK, so from the SQL results, the custom field "Server" has no choice set (and I see no choice set was built for it).

How exactly was this defined in NB 3.5?  If it was defined as type "Selection" then it should have had a comma-separated list of allowed choices, and this should have been migrated to a choice set. But if it was of type "Object" (e.g. pointing to a Device) then it should have remained as "Object" and not changed to "Selection".

**Reply by candlerb** [reply]
> do I need to revert back to a 3.5 series build for the mean time?

You could create the missing Choice Set and attach it to the 'Server' custom field, and all should then be OK.

But it would be helpful if you could revert to 3.5.x to understand what went wrong in the upgrade, and if this is reproducible and raised as an issue, then I'd expect the migration will be fixed.

**Reply by abhijitsharma1010** [reply]
I did the upgrade to v3.6 from v3.5.8 on my production vm as well as my testing vm. faced the same issue on both.

**Reply by ludof54** [reply]
Hello, i got the same type of error after an upgrade from v3.4.4 to 3.6 when i go to Admin / Current config 
![2023-09-06_10h10_19](https://github.com/netbox-community/netbox/assets/144216719/70dbc800-d87c-40db-abc3-495dd3586b2d)

**Reply by candlerb** [reply]
That's a different error. I think it's almost certainly issue #13622, which has been fixed in the 'develop' branch and should arrive in v3.6.1.

**Comment by abhijitsharma1010** [comment]
I upgraded to v3.6.1 and still i am getting this error. 
![14](https://github.com/netbox-community/netbox/assets/44794851/7c72e232-fc71-4498-90b1-bde82e8ba5e7)

**Reply by abhijitsharma1010** [reply]
Environment:


Request Method: GET
Request URL: https://10.40.122.145/virtualization/virtual-machines/

Django Version: 4.2.5
Python Version: 3.8.10
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.humanize',
 'corsheaders',
 'debug_toolbar',
 'graphiql_debug_toolbar',
 'django_filters',
 'django_tables2',
 'django_prometheus',
 'graphene_django',
 'mptt',
 'rest_framework',
 'social_django',
 'taggit',
 'timezone_field',
 'core',
 'account',
 'circuits',
 'dcim',
 'ipam',
 'extras',
 'tenancy',
 'users',
 'utilities',
 'virtualization',
 'wireless',
 'django_rq',
 'drf_spectacular',
 'drf_spectacular_sidecar']
Installed Middleware:
['graphiql_debug_toolbar.middleware.DebugToolbarMiddleware',
 'django_prometheus.middleware.PrometheusBeforeMiddleware',
 'corsheaders.middleware.CorsMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'django.middleware.security.SecurityMiddleware',
 'netbox.middleware.RemoteUserMiddleware',
 'netbox.middleware.CoreMiddleware',
 'netbox.middleware.MaintenanceModeMiddleware',
 'django_prometheus.middleware.PrometheusAfterMiddleware']



Traceback (most recent call last):
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
    response = get_response(request)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 104, in view
    return self.dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/base.py", line 77, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/netbox/utilities/views.py", line 99, in dispatch
    return super().dispatch(request, *args, **kwargs)
  File "/opt/netbox/venv/lib/python3.8/site-packages/django/views/generic/base.py", line 143, in dispatch
    return handler(request, *args, **kwargs)
  File "/opt/netbox/netbox/netbox/views/generic/bulk_views.py", line 179, in get
    'filter_form': self.filterset_form(request.GET, label_suffix='') if self.filterset_form else None,
  File "/opt/netbox/netbox/netbox/forms/base.py", line 166, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/netbox/netbox/utilities/forms/mixins.py", line 19, in __init__
    super().__init__(*args, **kwargs)
  File "/opt/netbox/netbox/extras/forms/mixins.py", line 30, in __init__
    self._append_customfield_fields()
  File "/opt/netbox/netbox/extras/forms/mixins.py", line 57, in _append_customfield_fields
    self.fields[field_name] = self._get_form_field(customfield)
  File "/opt/netbox/netbox/netbox/forms/base.py", line 181, in _get_form_field
    return customfield.to_form_field(set_initial=False, enforce_required=False, enforce_visibility=False)
  File "/opt/netbox/netbox/extras/models/customfields.py", line 426, in to_form_field
    choices = self.choice_set.choices

Exception Type: AttributeError at /virtualization/virtual-machines/
Exception Value: 'NoneType' object has no attribute 'choices'

**Reply by candlerb** [reply]
Correct. #13622 (`'NoneType' object has no attribute '_meta'`) is fixed by v3.6.1, but that's a different issue.

As explained earlier, your error is due to bad state in your database. To fix it, you'll need to create the missing ChoiceSet to your selection custom field.

Unless you can find a reproducible way to get your system into the state it's in, which you raise as a bug report, it's unlikely to be fixed in Netbox, since (so far) you're the only one to have observed it.

**Reply by abhijitsharma1010** [reply]
Hey @candlerb can you guide me on how can I rectify this issue in my setup.

**Reply by candlerb** [reply]
Choice sets are managed via the GUI, under Customization > Custom Field Choices. Once you have created one that matches your needs (I believe this is for your custom field called "Server"), go to Customization > Custom Fields and assign the choice set you've created to the custom field.
